### DISCUSSION TITLE
handleSubmit does not get input value from form

### DISCUSSION THREAD

**Discussion by TOcvfan** [react-hook-form/react-hook-form#7866] [Q&A] [answered]
Hi
I can't get my form to work, when I press the register button, it only gets the default value I have tried different ways of setting handleSubmit but I can't get it to work

[I have put the code in codesandbox here](https://codesandbox.io/s/test-5j3tyg?file=/src/App.js)

I hope someone can help

**Comment by Moshyfawn** [comment, marked as answer]
You pass down the handlers and the value
```jsx
<Text
  onChange={onChange}
  onBlur={onBlur}
  selected={value}
/>
```
but your `Text` component doesn't accept and doesn't assign any of them to the actual `OutlinedInput`'s callbacks
```jsx
      <OutlinedInput
        error={err}
        type={type}
        defaultValue={defaultV}
        endAdornment={
          <InputAdornment position="end">
            <IconButton aria-label="toggle password visibility" edge="end">
              {label === "Navn" ? (
                <AccountBoxOutlinedIcon />
              ) : (
                <EmailOutlinedIcon />
              )}
            </IconButton>
          </InputAdornment>
        }
        classes={outlinedInputClasses}
        label={err ? errors.message : label}
      />
```
Try assigning the field handlers and value in your `Text` component you are passing down in the form
```jsx
const Text = ({ onChange }) => (
  <OutlinedInput onChange={event => onChange(event.target.value)} />
)
```

**Reply by TOcvfan** [reply]
I have done as you said but I can't get the selector to work, all I get is undefined

**Reply by Moshyfawn** [reply]
You're over-complicating your form and getting wrong values.

You don't need to set up external states and set values on Submit button click.
The data _(form values)_ are passed from handle Submit to the callback you're providing it. No need to pass external state values to it.

- Omit `setValues` call on Submit button click
- Don't pass external state values to `onSubmitSignIn`
```jsx
<form onSubmit={handleSubmit(onSubmit)} ></form>
```

For `Selector` you're creating a race condition when setting state and trying to update the form state with it. Instead of doing
```jsx
onChange={(e) => {
  setValue(e.target.value);
  onChange(value);
}}
```
just pass the form field's `onChange` callback. You don't need another `const [value, setValue] = useState("")`

Please, refer to [React Hook Form docs](https://react-hook-form.com/api/usecontroller/controller) for examples and information on how to use it. It seems like you're going against the library instead of using it ;)
